File: quicksort.py
# basic quicksort algorithm for arbitrary array assuming build-in compare func


def main():
    arr = ['a','l','u', 'b', 'a','c','d']
    print(doQuickSort(arr))

def doQuickSort(arrIn):
    if len(arrIn) < 2:
        return arrIn
    pivot = arrIn[0]
    less = []
    greater = []
    for el in arrIn[1:]:
        if el < pivot:
            less.append(el)
        else:
            greater.append(el)
    # normal order: 
    return doQuickSort(less) +  [pivot] + doQuickSort(greater) 
    # reverse order:
    # return doQuickSort(greater) + [pivot] + doQuickSort(less) 



if __name__ == '__main__':
    main()

File: js/quicksort.js
const arr = [6, 3, 67, 9, 4, 5, 7, 8];

const qSort = function (arrIn) {
    if (arrIn.length < 2) {
        return arrIn;
    } else {
        const pivot = arrIn.shift();
        const less = arrIn.reduce((accum, el) => { if (el <= pivot) { accum.push(el); } return accum; }, []);
        const more = arrIn.reduce((accum, el) => { if (el > pivot) { accum.push(el); } return accum; }, []);

        return qSort(less).concat(pivot).concat(qSort(more));
    }
}

console.log(qSort(arr));

